File: concepts/03_errors_and_resources/01_resource_cleanup/julia/test_02_partial_acquisition_and_control_flow.jl
# polyglot-family: errors_and_resources
# polyglot-concept: resource_cleanup
# polyglot-related: languages/julia/language/07_unicode_exceptions_and_resources/
# polyglot-related+: test_054_try_catch_finally_and_rethrow.jl
#
# 共同问题：部分资源取得后失败怎样回滚；return 是否跳过 cleanup。
# 对照观察：finally 在 return 和 throw 前执行；只清理已经成功取得且由当前作用域拥有的资源。

using Test

function controlled_exit(log, fail)
    acquired = false
    try
        push!(log, :acquire)
        acquired = true
        fail && error("after acquire")
        return :done
    finally
        acquired && push!(log, :release)
    end
end

@testset "finally 保持 acquisition/release 配对" begin
    normal = Symbol[]
    @test controlled_exit(normal, false) == :done
    @test normal == [:acquire, :release]
    failing = Symbol[]
    @test_throws ErrorException controlled_exit(failing, true)
    @test failing == [:acquire, :release]
end

File: concepts/03_errors_and_resources/01_resource_cleanup/rust/test_02_partial_acquisition_and_control_flow.rs
// polyglot-family: errors_and_resources
// polyglot-concept: resource_cleanup
// polyglot-related: languages/rust/tests/language/08_errors_panics_and_resource_management/
// polyglot-related+: test_064_cleanup_failure_and_original_error.rs
//
// 共同问题：部分取得资源后失败怎样收尾；清理错误与原始错误冲突时保留什么。
// 对照观察：只为已构造的 guard 执行 Drop；可能失败的显式 close 必须由 API 组合多个错误。

#[derive(Debug, PartialEq)]
struct Failures {
    primary: &'static str,
    cleanup: &'static str,
}

#[test]
fn comparison() {
    let primary = Err::<(), _>("acquire second failed");
    let cleanup = Err::<(), _>("close first failed");
    let combined = match (primary, cleanup) {
        (Err(primary), Err(cleanup)) => Err(Failures { primary, cleanup }),
        _ => Ok(()),
    };
    assert_eq!(
        combined,
        Err(Failures {
            primary: "acquire second failed",
            cleanup: "close first failed"
        })
    );
}
